src/mame/jaleco/megasys1.cpp

```diff
@@ -5238,105 +5274,44 @@ void megasys1_bc_iosim_state::init_cybattlr()
+	m_ip_select_values[4] = 0x54;
 
-	m_ip_select_values[5] = 0xf0;
+	m_ip_select_values[5] = 0xfa;
 	m_ip_select_values[6] = 0x06;
 
 	save_item(NAME(m_ip_latched));
+	save_item(NAME(m_sprite_bank));
 }
 
-void megasys1_state::init_edfp()
-{
-	phantasm_rom_decode(machine(), "maincpu");
-}
-
-void megasys1_bc_iosim_state::init_hayaosi1()
+void megasys1_bc_iosim_state::init_cybattlr() // Type C
 {
-	m_ip_select_values[0] = 0x51;
+	m_ip_select_values[0] = 0x56;
 	m_ip_select_values[1] = 0x52;
 	m_ip_select_values[2] = 0x53;
 	m_ip_select_values[3] = 0x54;
 	m_ip_select_values[4] = 0x55;
 
-	m_ip_select_values[5] = 0xfc;
+	m_ip_select_values[5] = 0xf2;
 	m_ip_select_values[6] = 0x06;
 
 	save_item(NAME(m_ip_latched));
+	save_item(NAME(m_sprite_bank));
 }
 
-u16 megasys1_state::iganinju_mcu_hs_r(offs_t offset, u16 mem_mask)
-{
-	if(m_mcu_hs && ((m_mcu_hs_ram[8/2] << 6) & 0x3ffc0) == ((offset*2) & 0x3ffc0))
-	{
-		if(MCU_HS_LOG && !machine().side_effects_disabled())
-			printf("MCU HS R (%04x) <- [%02x]\n",mem_mask,offset*2);
-
-		return 0x835d;
-	}
-
-	return m_rom_maincpu[offset];
-}
-
-void megasys1_state::iganinju_mcu_hs_w(offs_t offset, u16 data, u16 mem_mask)
-{
-	// [0/2]: 0x0000
-	// [2/2]: 0x0055
-	// [4/2]: 0x00aa
-	// [6/2]: 0x00ff
-	// [8/2]: 0x0bc0
-	// expects 0x835d to be read at 0x2f000, does hs sequence until that happens
-
-	COMBINE_DATA(&m_mcu_hs_ram[offset]);
-
-	if(MCU_HS_SEQ(0x0000,0x0055,0x00aa,0x00ff) && offset == 0x8/2)
-		m_mcu_hs = 1;
-	else
-		m_mcu_hs = 0;
-
-	if(MCU_HS_LOG && !machine().side_effects_disabled())
-		printf("MCU HS W %04x (%04x) -> [%02x]\n",data,mem_mask,offset*2);
-}
-
-void megasys1_state::init_iganinju()
-{
-	phantasm_rom_decode(machine(), "maincpu");
-
-	m_maincpu->space(AS_PROGRAM).install_read_handler(0x00000, 0x3ffff, read16s_delegate(*this, FUNC(megasys1_state::iganinju_mcu_hs_r)));
-	m_maincpu->space(AS_PROGRAM).install_write_handler(0x2f000, 0x2f009, write16s_delegate(*this, FUNC(megasys1_state::iganinju_mcu_hs_w)));
-
-	//m_rom_maincpu[0x00006e/2] = 0x0420; // the only game that does
-										// not like lev 3 interrupts
-
-	m_mcu_hs = 0;
-	memset(m_mcu_hs_ram, 0, sizeof(m_mcu_hs_ram));
-
-	save_item(NAME(m_mcu_hs));
-	save_item(NAME(m_mcu_hs_ram));
-}
-
-void megasys1_state::init_jitsupro()
+void megasys1_bc_iomcu_state::init_bigstrik() // Type C
 {
-	astyanax_rom_decode(machine(), "maincpu");      // Code
-
-	jitsupro_gfx_unmangle("scroll0");   // Gfx
-	jitsupro_gfx_unmangle("sprites");
-	m_maincpu->space(AS_PROGRAM).install_read_handler(0x00000, 0x3ffff, read16s_delegate(*this, FUNC(megasys1_state::megasys1A_mcu_hs_r)));
-	m_maincpu->space(AS_PROGRAM).install_write_handler(0x20000, 0x20009, write16s_delegate(*this, FUNC(megasys1_state::megasys1A_mcu_hs_w)));
-
-	m_mcu_hs = 0;
-	memset(m_mcu_hs_ram, 0, sizeof(m_mcu_hs_ram));
-
-	save_item(NAME(m_mcu_hs));
-	save_item(NAME(m_mcu_hs_ram));
+	save_item(NAME(m_sprite_bank));
+	save_item(NAME(m_mcu_input_data));
+	save_item(NAME(m_mcu_io_data));
 }
```

Comment:
Does this really need an init function?  Can’t the class with the emulated microcontroller just override `machine_start` to register things for save states?

Reply:
it now only has the sprite_bank bit as it's type c, the rest is in a machine_start